e2e_tests/tests/cluster/test_agent_restart.py

```diff
@@ -0,0 +1,222 @@
+import json
+import os
+import subprocess
+import time
+from typing import Any, Dict, Iterator, List, Tuple, cast
+
+import pytest
+
+from tests import config as conf
+from tests import experiment as exp
+
+from .utils import get_command_info, run_command, wait_for_command_state
+
+DEVCLUSTER_CONFIG_PATH = conf.PROJECT_ROOT_PATH.joinpath(
+    ".circleci/devcluster/double.devcluster.yaml"
+)
+
+
+def _get_agent_data(master_url: str) -> List[Dict[str, Any]]:
+    command = ["det", "-m", master_url, "agent", "list", "--json"]
+    output = subprocess.check_output(command).decode()
+    agent_data = cast(List[Dict[str, Any]], json.loads(output))
+    return agent_data
+
+
+class ManagedCluster:
+    def __init__(self) -> None:
+        # Strategically only import devcluster on demand to avoid having it as a hard dependency.
+        from devcluster import Devcluster
+
+        self.dc = Devcluster(config=str(DEVCLUSTER_CONFIG_PATH))
+        self.master_url = conf.make_master_url()
+        self.old_cd = os.getcwd()
+
+    def __enter__(self) -> "ManagedCluster":
+        self.old_cd = os.getcwd()
+        os.chdir(str(conf.PROJECT_ROOT_PATH))
+        self.dc.__enter__()
+        return self
+
+    def __exit__(self, *_: Any) -> None:
+        os.chdir(self.old_cd)
+        self.dc.__exit__(*_)
+
+    def initial_startup(self) -> None:
+        self.dc.set_target("agent1", wait=True, timeout=3 * 60)
+
+    def kill_agent(self) -> None:
+        self.dc.kill_stage("agent1")
+
+        WAIT_FOR_KILL = 5
+        for _i in range(WAIT_FOR_KILL):
+            agent_data = _get_agent_data(self.master_url)
+            if len(agent_data) == 0:
+                break
+            if len(agent_data) == 1 and agent_data[0]["draining"] is True:
+                break
+            time.sleep(1)
+        else:
+            pytest.fail(f"Agent is still present after {WAIT_FOR_KILL} seconds")
+
+    def restart_agent(self) -> None:
+        agent_data = _get_agent_data(self.master_url)
+        if len(agent_data) == 1 and agent_data[0]["enabled"]:
+            return
+
+        # Currently, we've got to wait for master to "forget" the agent before reconnecting.
+        WAIT_FOR_AMNESIA = 60
+        for _i in range(WAIT_FOR_AMNESIA):
+            agent_data = _get_agent_data(self.master_url)
```

Comment:
it's a bit awkward to read this when it uses double then only expects to have and lose one agent when it uses `double.devcluster.yaml`, but then i noticed we only ever go to stage `agent1`.  just a comment in init or something could help.